File: backend/app/api/public.py
"""Anonymous read API. Serves only the public projection (Publication rows) plus atlas structure.
Nothing here can launch work or reveal private records."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from ..db import get_db
from ..models import Area, Event, Publication, Relation

router = APIRouter(prefix="/public", tags=["public"])


def _latest_publications(db: Session, record_type: str | None = None) -> list[Publication]:
    """Latest non-withdrawn version per record."""
    stmt = select(Publication).where(Publication.withdrawn_at.is_(None))
    if record_type:
        stmt = stmt.where(Publication.record_type == record_type)
    rows = db.scalars(stmt.order_by(Publication.published_at)).all()
    latest: dict[tuple[str, str], Publication] = {}
    for row in rows:
        latest[(row.record_type, row.record_id)] = row
    return list(latest.values())


def _serialize(pub: Publication) -> dict:
    return {
        "record_type": pub.record_type,
        "record_id": pub.record_id,
        "version": pub.record_version,
        "evidence_label": pub.evidence_label,
        "policy_version": pub.policy_version,
        "published_at": pub.published_at.isoformat(),
        **pub.public_payload,
    }


@router.get("/areas")
def list_areas(db: Session = Depends(get_db)) -> list[dict]:
    areas = db.scalars(select(Area).order_by(Area.depth, Area.name)).all()
    counts: dict[str, int] = {
        target_id: n
        for target_id, n in db.execute(
            select(Relation.target_id, func.count())
            .where(Relation.layer == "atlas", Relation.kind == "classified_in")
            .group_by(Relation.target_id)
        ).all()
    }
    return [
        {
            "id": a.id,
            "slug": a.slug,
            "name": a.name,
            "description": a.description,
            "parent_id": a.parent_id,
            "depth": a.depth,
            "problem_count": counts.get(a.id, 0),
        }
        for a in areas
    ]


@router.get("/problems")
def list_problems(area: str | None = None, db: Session = Depends(get_db)) -> list[dict]:
    pubs = [_serialize(p) for p in _latest_publications(db, "problem")]
    if area:
        pubs = [p for p in pubs if any(a["slug"] == area for a in p.get("areas", []))]
    return sorted(pubs, key=lambda p: p["title"])


@router.get("/problems/{slug}")
def get_problem(slug: str, db: Session = Depends(get_db)) -> dict:
    for pub in _latest_publications(db, "problem"):
        if pub.public_payload.get("slug") == slug:
            problem = _serialize(pub)
            problem_id = pub.record_id
            campaigns = [
                _serialize(c)
                for c in _latest_publications(db, "campaign")
                if c.public_payload.get("problem_id") == problem_id
            ]
            ideas = [
                _serialize(i)
                for i in _latest_publications(db, "idea")
                if i.public_payload.get("problem_id") == problem_id
            ]
            idea_ids = {i["id"] for i in ideas}
            claims = [
                _serialize(c)
                for c in _latest_publications(db, "claim")
                if c.public_payload.get("idea_id") in idea_ids
            ]
            evidence = [
                _serialize(e)
                for e in _latest_publications(db, "evidence")
                if e.public_payload.get("idea_id") in idea_ids
                or e.public_payload.get("problem_id") == problem_id
            ]
            return {
                "problem": problem,
                "campaigns": campaigns,
                "ideas": ideas,
                "claims": claims,
                "evidence": evidence,
            }
    raise HTTPException(404, "No published problem with that slug")


@router.get("/ideas/{idea_id}")
def get_idea(idea_id: str, db: Session = Depends(get_db)) -> dict:
    idea = next((p for p in _latest_publications(db, "idea") if p.record_id == idea_id), None)
    if idea is None:
        raise HTTPException(404, "No published idea with that id")
    claims = [
        _serialize(c)
        for c in _latest_publications(db, "claim")
        if c.public_payload.get("idea_id") == idea_id
    ]
    evidence = [
        _serialize(e)
        for e in _latest_publications(db, "evidence")
        if e.public_payload.get("idea_id") == idea_id
    ]
    return {"idea": _serialize(idea), "claims": claims, "evidence": evidence}


@router.get("/graph")
def graph(
    layers: str = Query("atlas,lineage,dependency,association"),
    problem: str | None = None,
    db: Session = Depends(get_db),
) -> dict:
    """Four-layer graph restricted to published records. Geometry is chosen by the client; it is
    never a mathematical distance (docs/plan.md §8.2)."""
    wanted = set(layers.split(","))
    nodes: dict[str, dict] = {}
    areas = db.scalars(select(Area)).all()
    if "atlas" in wanted:
        for a in areas:
            nodes[a.id] = {
                "id": a.id,
                "type": "area",
                "label": a.name,
                "slug": a.slug,
                "depth": a.depth,
                "parent_id": a.parent_id,
            }
    problems = _latest_publications(db, "problem")
    if problem:
        problems = [p for p in problems if p.public_payload.get("slug") == problem]
    problem_ids = {p.record_id for p in problems}
    for p in problems:
        nodes[p.record_id] = {
            "id": p.record_id,
            "type": "problem",
            "label": p.public_payload["title"],
            "slug": p.public_payload["slug"],
            "status": p.public_payload["status"],
            "evidence_label": p.evidence_label,
            "areas": [a["slug"] for a in p.public_payload.get("areas", [])],
        }
    ideas = [
        i
        for i in _latest_publications(db, "idea")
        if i.public_payload.get("problem_id") in problem_ids
    ]
    for i in ideas:
        pl = i.public_payload
        nodes[i.record_id] = {
            "id": i.record_id,
            "type": "idea",
            "label": pl["title"],
            "problem_id": pl["problem_id"],
            "generation": pl["generation"],
            "depth": pl["depth"],
            "evidence_status": pl["evidence_status"],
            "scheduling_status": pl["scheduling_status"],
            "formalization_status": pl["formalization_status"],
            "evidence_label": i.evidence_label,
            "score": pl["score"],
            "method_tags": pl["method_tags"],
        }
    idea_ids = {i.record_id for i in ideas}
    claims = [
        c for c in _latest_publications(db, "claim") if c.public_payload.get("idea_id") in idea_ids
    ]
    for c in claims:
        nodes[c.record_id] = {
            "id": c.record_id,
            "type": "claim",
            "label": c.public_payload["statement"][:80],
            "idea_id": c.public_payload["idea_id"],
            "evidence_label": c.evidence_label,
            "formalization_status": c.public_payload["formalization_status"],
        }

    links: list[dict] = []
    if "atlas" in wanted:
        for a in areas:
            if a.parent_id and a.parent_id in nodes:
                links.append(
                    {"source": a.id, "target": a.parent_id, "layer": "atlas", "kind": "subfield_of"}
                )
    relations = db.scalars(select(Relation).where(Relation.layer.in_(wanted))).all()
    for r in relations:
        if r.source_id in nodes and r.target_id in nodes:
            links.append(
                {
                    "source": r.source_id,
                    "target": r.target_id,
                    "layer": r.layer,
                    "kind": r.kind,
                    "status": r.status,
                }
            )
    return {"nodes": list(nodes.values()), "links": links, "layers": sorted(wanted)}


@router.get("/events")
def public_events(
    since_id: int = 0, limit: int = Query(200, le=1000), db: Session = Depends(get_db)
) -> list[dict]:
    """Public timeline for replay. Only events written with visibility=public are visible; their
    payloads are constructed without private fields."""
    rows = db.scalars(
        select(Event)
        .where(Event.visibility == "public", Event.id > since_id)
        .order_by(Event.id)
        .limit(limit)
    ).all()
    return [
        {
            "id": e.id,
            "type": e.type,
            "record_type": e.record_type,
            "record_id": e.record_id,
            "payload": e.payload,
            "created_at": e.created_at.isoformat(),
        }
        for e in rows
    ]


@router.get("/labels")
def labels() -> dict:
    from ..services.publication import EVIDENCE_LABELS, PROBLEM_STATUS_LABELS

    return {"evidence": EVIDENCE_LABELS, "problem_status": PROBLEM_STATUS_LABELS}

File: apps/research-api/src/public.ts
// Anonymous read API for the published atlas — port of MathLab's public.py.
// Serves only the public projection (publication rows) plus atlas structure;
// nothing here can launch work or reveal private records.

import type { FastifyInstance, FastifyRequest } from "fastify";
import {
  EVIDENCE_LABELS,
  PROBLEM_STATUS_LABELS,
  getCollections,
  latestPublications,
  serializePublication,
} from "@triviality/database";
import type { PublicationDocument } from "@triviality/database";

function serializeAll(pubs: PublicationDocument[]): Record<string, unknown>[] {
  return pubs.map(serializePublication);
}

export function registerPublicRoutes(app: FastifyInstance): void {
  app.get("/public/areas", async () => {
    const collections = await getCollections();
    const [areas, counts] = await Promise.all([
      collections.atlasAreas.find({}).sort({ depth: 1, name: 1 }).toArray(),
      collections.relations
        .aggregate<{ _id: string; n: number }>([
          { $match: { layer: "atlas", kind: "classified_in" } },
          { $group: { _id: "$targetId", n: { $sum: 1 } } },
        ])
        .toArray(),
    ]);
    const countByArea = new Map(counts.map((c) => [c._id, c.n]));
    return areas.map((a) => ({
      id: a._id,
      slug: a.slug,
      name: a.name,
      description: a.description ?? "",
      parent_id: a.parentId,
      depth: a.depth,
      problem_count: countByArea.get(a._id) ?? 0,
    }));
  });

  app.get(
    "/public/problems",
    async (request: FastifyRequest<{ Querystring: { area?: string } }>) => {
      const collections = await getCollections();
      let pubs = serializeAll(await latestPublications(collections, "problem"));
      const area = request.query.area;
      if (area) {
        pubs = pubs.filter((p) =>
          (p.areas as { slug: string }[] | undefined)?.some((a) => a.slug === area),
        );
      }
      return pubs.sort((a, b) => String(a.title).localeCompare(String(b.title)));
    },
  );

  app.get(
    "/public/problems/:slug",
    async (request: FastifyRequest<{ Params: { slug: string } }>, reply) => {
      const collections = await getCollections();
      const problems = await latestPublications(collections, "problem");
      const pub = problems.find((p) => p.publicPayload.slug === request.params.slug);
      if (!pub) return reply.code(404).send({ error: "No published problem with that slug" });
      const problemId = pub.recordId;

      const [campaignPubs, ideaPubs, claimPubs, evidencePubs] = await Promise.all([
        latestPublications(collections, "campaign"),
        latestPublications(collections, "idea"),
        latestPublications(collections, "claim"),
        latestPublications(collections, "evidence"),
      ]);
      const campaigns = campaignPubs.filter((c) => c.publicPayload.problem_id === problemId);
      const ideas = ideaPubs.filter((i) => i.publicPayload.problem_id === problemId);
      const ideaIds = new Set(ideas.map((i) => i.recordId));
      const claims = claimPubs.filter(
        (c) => ideaIds.has(String(c.publicPayload.idea_id)) || c.publicPayload.problem_id === problemId,
      );
      const evidence = evidencePubs.filter(
        (e) =>
          ideaIds.has(String(e.publicPayload.idea_id)) ||
          e.publicPayload.problem_id === problemId,
      );
      return {
        problem: serializePublication(pub),
        campaigns: serializeAll(campaigns),
        ideas: serializeAll(ideas),
        claims: serializeAll(claims),
        evidence: serializeAll(evidence),
      };
    },
  );

  app.get(
    "/public/ideas/:id",
    async (request: FastifyRequest<{ Params: { id: string } }>, reply) => {
      const collections = await getCollections();
      const idea = (await latestPublications(collections, "idea")).find(
        (p) => p.recordId === request.params.id,
      );
      if (!idea) return reply.code(404).send({ error: "No published idea with that id" });
      const [claimPubs, evidencePubs] = await Promise.all([
        latestPublications(collections, "claim"),
        latestPublications(collections, "evidence"),
      ]);
      return {
        idea: serializePublication(idea),
        claims: serializeAll(
          claimPubs.filter((c) => c.publicPayload.idea_id === idea.recordId),
        ),
        evidence: serializeAll(
          evidencePubs.filter((e) => e.publicPayload.idea_id === idea.recordId),
        ),
      };
    },
  );

  app.get(
    "/public/graph",
    async (
      request: FastifyRequest<{ Querystring: { layers?: string; problem?: string } }>,
    ) => {
      const collections = await getCollections();
      const wanted = new Set(
        (request.query.layers ?? "atlas,lineage,dependency,association").split(","),
      );
      const nodes = new Map<string, Record<string, unknown>>();
      const areas = await collections.atlasAreas.find({}).toArray();
      if (wanted.has("atlas")) {
        for (const a of areas) {
          nodes.set(a._id, {
            id: a._id,
            type: "area",
            label: a.name,
            slug: a.slug,
            depth: a.depth,
            parent_id: a.parentId,
          });
        }
      }

      let problems = await latestPublications(collections, "problem");
      const problemSlug = request.query.problem;
      if (problemSlug) problems = problems.filter((p) => p.publicPayload.slug === problemSlug);
      const problemIds = new Set(problems.map((p) => p.recordId));
      for (const p of problems) {
        nodes.set(p.recordId, {
          id: p.recordId,
          type: "problem",
          label: p.publicPayload.title,
          slug: p.publicPayload.slug,
          status: p.publicPayload.status,
          evidence_label: p.evidenceLabel,
          areas: (p.publicPayload.areas as { slug: string }[] | undefined)?.map((a) => a.slug) ?? [],
        });
      }

      const ideas = (await latestPublications(collections, "idea")).filter((i) =>
        problemIds.has(String(i.publicPayload.problem_id)),
      );
      for (const i of ideas) {
        const pl = i.publicPayload;
        nodes.set(i.recordId, {
          id: i.recordId,
          type: "idea",
          label: pl.title,
          problem_id: pl.problem_id,
          generation: pl.generation,
          depth: pl.depth,
          evidence_status: pl.evidence_status,
          scheduling_status: pl.scheduling_status,
          formalization_status: pl.formalization_status,
          evidence_label: i.evidenceLabel,
          score: pl.score,
          method_tags: pl.method_tags,
        });
      }
      const ideaIds = new Set(ideas.map((i) => i.recordId));
      const claims = (await latestPublications(collections, "claim")).filter((c) =>
        ideaIds.has(String(c.publicPayload.idea_id)) || problemIds.has(String(c.publicPayload.problem_id)),
      );
      for (const c of claims) {
        nodes.set(c.recordId, {
          id: c.recordId,
          type: "claim",
          label: String(c.publicPayload.statement ?? "").slice(0, 80),
          idea_id: c.publicPayload.idea_id,
          evidence_label: c.evidenceLabel,
          formalization_status: c.publicPayload.formalization_status,
        });
      }

      const links: Record<string, unknown>[] = [];
      if (wanted.has("atlas")) {
        for (const a of areas) {
          if (a.parentId && nodes.has(a.parentId)) {
            links.push({ source: a._id, target: a.parentId, layer: "atlas", kind: "subfield_of" });
          }
        }
      }
      const relations = await collections.relations
        .find({ layer: { $in: [...wanted] } })
        .toArray();
      for (const r of relations) {
        if (nodes.has(r.sourceId) && nodes.has(r.targetId)) {
          links.push({
            source: r.sourceId,
            target: r.targetId,
            layer: r.layer,
            kind: r.kind,
            status: r.status,
          });
        }
      }
      return { nodes: [...nodes.values()], links, layers: [...wanted].sort() };
    },
  );

  app.get(
    "/public/events",
    async (
      request: FastifyRequest<{ Querystring: { since_id?: string; limit?: string } }>,
    ) => {
      const collections = await getCollections();
      const sinceId = Math.max(0, Number(request.query.since_id ?? 0) || 0);
      const limit = Math.min(1000, Math.max(1, Number(request.query.limit ?? 200) || 200));
      const rows = await collections.publicEvents
        .find({ seq: { $gt: sinceId } })
        .sort({ seq: 1 })
        .limit(limit)
        .toArray();
      return rows.map((e) => ({
        id: e.seq,
        type: e.type,
        record_type: e.recordType,
        record_id: e.recordId,
        payload: e.payload,
        created_at: e.createdAt.toISOString(),
      }));
    },
  );

  app.get("/public/labels", async () => ({
    evidence: EVIDENCE_LABELS,
    problem_status: PROBLEM_STATUS_LABELS,
  }));
}
